### ISSUE TITLE
[ feature ] Show in-progress preview in the `Image Preview` node

### ISSUE THREAD

**Issue by JorgeR81** [chrisgoringe/cg-controller#263]
If an `Image Preview` node is added to the Controller, it would show the `KSampler` in-progress preview

This could be useful for people that want to build a whole GUI, like SwarmUI or InvokeAI.
The "big picture area" can also show the in-progress preview.

---

There is a workaround for this, but it may be too complex for new users:

In Comfy UI, you can create a group node with: `KSampler` + `Vae Decode` + `Image Preview`.
This way, the final image will show up in the same node as the in progress preview.

You can use the `SamplerCustomAdvanced` node, so that there are no other widgets.
The you can add the group node the the Controller.

**Comment by LukeG89** [comment]
I think that the preview is only available on sampler nodes, you can associate a preview node to a ksampler latent preview (I guess).

Maybe a future workaroud would be through the "hide widgets" feature created by Pythongosssss.

**Comment by chrisgoringe** [comment]
It's easy enough to catch the preview image - we already do it, in fact, in order to show the preview even if the KSampler node is minimised (minimised nodes don't get images for obvious reasons). So showing the preview wouldn't be technically hard.

Wondering what the UI would be like, though - a workflow might have multiple Image Preview nodes, (and/or multiple KSamplers) and it wouldn't make sense to show all KSampler in progress previews on all Image Previews.

It could be done if they are in the same group, perhaps, or if they are directly connected to each other in the workflow? Or would we need a way for the user to specify?

I'm leaning to the same group (and maybe only if there is exactly one preview and exactly one sampler)

**Comment by JorgeR81** [comment]
> Wondering what the UI would be like, though - a workflow might have multiple Image Preview nodes, (and/or multiple KSamplers) and it wouldn't make sense to show all KSampler in progress previews on all Image Previews.

You would only show in in-progress preview, in the `Image Preview` nodes that are in the Controller.
You would place the in-progress preview in the same exact area were the final image will show up.
And once the generation is done, the in-progress preview would be replaced by the final image.

The main use case is a layout like this, with a single image preview. 
I want to see the preview, while the image is being generated, like in a traditional GUI.

![p1b](https://github.com/user-attachments/assets/9cd69236-a230-4eba-b386-2068c2a0685c)

**Comment by LukeG89** [comment]
> or if they are directly connected to each other in the workflow?

I would choose this option, to associate the preview node to the KSampler that is generating the image.

**Comment by JorgeR81** [comment]
> > or if they are directly connected to each other in the workflow?
> 
> I would choose this option, to associate the preview node to the KSampler that is generating the image.

Yes, exactly.

**Comment by JorgeR81** [comment]
> I'm using the rgthree progress bar it right now!
> 
> I don't think we need to replicate this functionality.


And ( unlike I suggested before ) we actually could use a progress bar  for the controller.
 
It could be shown below the in progress preview !

**Comment by chrisgoringe** [comment]
> > or if they are directly connected to each other in the workflow?
> 
> 
> 
> I would choose this option, to associate the preview node to the KSampler that is generating the image.

Actually, it's not directly connected, there is a VAE decode in between.

**Comment by LukeG89** [comment]
> Actually, it's not directly connected, there is a VAE decode in between.

@chrisgoringe Damn it! 😆 

> It could be done if they are in the same group

But this way you will have both KSampler and Preview Image displaying the latent preview.

**Comment by chrisgoringe** [comment]
I think there are a few heuristics that could be used to link a preview node (or save image, for that matter) to a sampler. 

One would be to follow the dataflow upstream - what node did the image come from? What node sent that node either an image or a latent? etc until you hit a sampler.

**Comment by chrisgoringe** [comment]
Um... I think I might have made this happen...

**Comment by LukeG89** [comment]
It works, but it's glitchy...


https://github.com/user-attachments/assets/823fbbd6-cc3d-44ad-8540-67d2716b1ff1

**Comment by JorgeR81** [comment]
> It works, but it's glitchy...

It seems that between steps, it's shown the final image, that was previously generated.

**Comment by LukeG89** [comment]
It seems the glitch happens in certain zoom/canvas position (for example, it doesn't glitch if you are far from KSampler)

**Comment by JorgeR81** [comment]
There's also a different bug:

An image preview is showing up in the `Load VAE` node

![vae](https://github.com/user-attachments/assets/1a77d1fa-ab2f-4f46-97a5-e0ca0cbf19a1)

**Comment by JorgeR81** [comment]
It seems the glitch happens in certain zoom/canvas position (for example, it doesn't glitch if you are far from KSampler)

Yeah, with more "zoom in" and the Sampler out of the canvas, this does not happen.

**Comment by LukeG89** [comment]
@chrisgoringe Now the preview it's perfecly working in simple workflows with one sampler+preview, but with 2 pass + face detailer there are some issues:

- the 1st and 2nd pass previews are displayed in both 2pass preview and face detailer preview
- when in face detailer, the glitch is back
- after the generation, both previews in Controller show the latent previews and not the actual images


https://github.com/user-attachments/assets/d138bf74-a68a-46ac-9500-35d171d475b0

**Comment by chrisgoringe** [comment]
Yeah, this is where the “follow upstream” logic becomes less reliable!

One option would be to only show the preview if generated up to N links
upstream. N=2 would cover the basic case (sampler-vae decode-preview).

But if you do sampler-refiner-decode-preview, I think you would want the
preview to show the first and second samplers? That would be N=3…

I don’t do very complicated workflows so I don’t really have an instinct
for how it ought to work!

**Comment by chrisgoringe** [comment]
Could post a screenshot of the workflow corresponding to your video?

**Comment by LukeG89** [comment]
> I don’t do very complicated workflows so I don’t really have an instinct for how it ought to work!

Yeah, I just made one from scratch because I know ComfyUI users like complicated things! 😆 

> Could post a screenshot of the workflow corresponding to your video?

![workflow](https://github.com/user-attachments/assets/7a0afc5f-0a1d-42bc-bd0b-488927526add)

Here's the workflow in case you need
[1.3 - Group Tabs testing.json](https://github.com/user-attachments/files/17609811/1.3.-.Group.Tabs.testing.json)

**Comment by JorgeR81** [comment]
@chrisgoringe, Don't change this behavior, I love it. It's perfect !

The current "in progress" preview is also shown in the `Image Preview` node of the next steps !!!
Makes perfect sense. Because the next stages will be affected by changes in the previous ones.
 
And I can see all "in-progress previews", with a single image window open ( the one in the last stage ).
We can have only one big window for multiple stages, like a traditional GUI does.

---

If I make a change in the first Sampler. It shows the "in progress" preview in all windows.

![s1](https://github.com/user-attachments/assets/210a9f55-e213-4e18-a0b0-fd23617cf939)

---

If I make a change in the 2 stage ( refiner ), it only shows the  "in progress" preview in the refiner and the next step ( face detailer )

![s2](https://github.com/user-attachments/assets/93b2fe81-a373-4503-b298-d4f2d1823738)

---

And if I make a change in the last stage ( face detailer ) it shows the "in progress" preview only in the face detailer.

![s3](https://github.com/user-attachments/assets/21c7a2e5-f56e-4077-888b-7d6d69962716)

**Comment by JorgeR81** [comment]
But there are a couple of things that can be improved:

---

- When the generation is done, show the final image ( PNG ).
Now, if it do try to save the image via right click, I get a jfif file ( preview ), not the PNG

![prev](https://github.com/user-attachments/assets/743994fb-4c57-4edf-8b83-1fba77c01314)

---

- If a stage is bypassed, whe don't need to the the "in-progress" preview

![s1b](https://github.com/user-attachments/assets/33c6a4db-1129-4683-9de9-66a74f3fc71b)

**Comment by LukeG89** [comment]
@JorgeR81 Are you experiencing these bugs?
> - when in face detailer, the glitch is back
> - after the generation, both previews in Controller show the latent previews and not the actual images

> - after controlnet preprocessor
(no problem with that, except that it was previewing the Load Image connected to the preprocessor, but only before the generation)

**Comment by JorgeR81** [comment]
> - when in face detailer, the glitch is back
> - after the generation, both previews in Controller show the latent previews and not the actual images

Yes, I just updated my previous post with these issues.

I will check controlnet preprocessor.

**Comment by JorgeR81** [comment]
Except for the issues mentioned, it works as expected with all enabled.
**And it looks fantastic !**

But there are CPU spikes and ComfyUI slows down, when we try to move the canvas while generating.

![all](https://github.com/user-attachments/assets/e4234c67-94ba-4d4d-95a3-5c2cac8970ba)

**Comment by chrisgoringe** [comment]
Just pushed an update which fixes some of these:

- I think the final image glitch should be gone
- bypassed nodes don't get updated
- image moved above widgets

I'll look at Load Image next.

A question about the final image shown in the sampler node. Comfy leaves the sampler node showing the approximate image, not the finished one; that's currently what we're doing too. Personally, I'm ok with that - I think you can just watch the Image Preview node to see progress and final outcome.

The problem with "show the final image" is "which final image?". If there are multiple Image Preview nodes, which show different stages of the process, which image should the sampler show? 

I think it makes more sense for the sampler just to show the interim images, with the final node showing both.

**Comment by chrisgoringe** [comment]
OK, I think I've also fixed the issues with the Load Image node. LMK if it's still got problems.

**Comment by LukeG89** [comment]
Not the glitch is gone, but I still have this problem:

-  after the generation, both previews in Controller show the latent previews and not the actual images

At first I see the final image, but when I switch group, I see the latent preview.

![latent](https://github.com/user-attachments/assets/751eb31d-4d15-4dc1-a544-ffedf8cbb803)

____

Anyways, something happened to all the previews, now they are over the widgets (even in Load Image node)

![image](https://github.com/user-attachments/assets/e066778e-aede-4751-83a4-84f2b84f4b46)

**Comment by JorgeR81** [comment]
> - after the generation, both previews in Controller show the latent previews and not the actual images

> At first I see the final image, but when I switch group, I see the latent preview.

Same happens to me.

First I can get the PNG. 
But if I switch groups in the Controller I get the jfif file ( preview ).

---

@chrisgoringe, can you also hide the in-progress preview in the `Sampler`, if the Controller also as an `Image Preview`

Having the preview in the Sampler it's redundant now. It just takes space.

![b1x](https://github.com/user-attachments/assets/b520a304-02b0-44e6-b45b-7181cc0f6bd1)

**Comment by JorgeR81** [comment]
I can create a new group ( in blue ) just with an "Image loader", after the final stage.
And this "big picture area" shows all in-progress images and final previews, for the whole workflow, even with bypassed stages.

![bx2](https://github.com/user-attachments/assets/86bf21ae-9fa8-4e01-a922-8f03fc85c24c)

---

And I can collapse  the "big picture area" to see the full workflow.  

![bx1](https://github.com/user-attachments/assets/12ce4a8b-ef27-4284-b9e9-fba841814312)

**Comment by JorgeR81** [comment]
@chrisgoringe, about the progress bar, the rgthree bar works well, on top of the workflow.

What we could use in the Controller is that mini progress bar that shows up in the Sampler header. 

![bx3](https://github.com/user-attachments/assets/10342ab3-35bb-49cb-888e-34560442e23b)

**Comment by JorgeR81** [comment]
> @chrisgoringe, can you also hide the in-progress preview in the Sampler, if a Controller also has an Image Preview, displaying that same in-progress preview ?

Maybe we don't even need to hide the `Sampler` previews.

With this configuration, we don't need an `Image Preview` in the other Controllers. 
We can get the PNG file from the "Big Picture area".

![bx5](https://github.com/user-attachments/assets/1de6bef2-9b10-4837-a210-bcfc086f2b8e)

**Comment by JorgeR81** [comment]
With this more complex configuration, sometimes ComfyUI gets very slow.

Then sometimes, the whole ComfyUI page gets white, and I get an `Out of Memory` error code, right in the browser's page.
If I Reload / Refresh the browser, it starts working again, and it's fast again.

I use Google Chrome, in Windows 10, and I have 32 GB RAM.

---

![rel](https://github.com/user-attachments/assets/80a8c081-355d-4269-9e70-29404f8045fa)

**Comment by JorgeR81** [comment]
> @chrisgoringe, can you also hide the in-progress preview in the Sampler, if a Controller also has an Image Preview, displaying that same in-progress preview ?

A better solution would be to show the final PGN image, in the KSampler, once the generation is done.
This way we don't need to have an `Image Preview` node in the Controller !

Is that possible ?

![btr](https://github.com/user-attachments/assets/d1a2b0c3-6121-42fa-8e3c-dc70bfeae6e2)

**Comment by JorgeR81** [comment]
This is probably the more complex use case. 

I added 2 more Controllers:

- with a `Ultimate SD Upscale` node ( tiled upscaler )
- and finally and a regular upscaler ( with the `x1_ITF_SkinDiffDetail_Lite_v1` model to give some "texture" to the image ).

![7c](https://github.com/user-attachments/assets/b79017fe-cbb6-4ca9-aec4-cbb968194756)

**Comment by chrisgoringe** [comment]
> > @chrisgoringe, can you also hide the in-progress preview in the Sampler, if a Controller also has an Image Preview, displaying that same in-progress preview ?
> 
> 
> 
> A better solution would be to show the final PGN image, in the Sampler, once the generation is done.
> 
> This way we don't need to have an `Image Preview` node in the Controller !
> 

I think I mentioned this above. The problem is, which final image to show?

If you have two stages and an image preview after each, what should the first sampler show when it finishes?

You get the situation where the behaviour of a node depends on what is connected to it *downstream* (where are image previews?) which is contrary to the way node based programming works and would, I think, inevitably lead to counterintuitive outcomes.

Comfy leaves the final interim image in place.

**Comment by chrisgoringe** [comment]
I think there are two issues mentioned above that I'll look at:

The image preview node returning to the last interim image

Images "forgetting" their size

(and maybe performance?)

**Comment by JorgeR81** [comment]
> I think I mentioned this above. The problem is, which final image to show?

What about the first alternative?
Don't show a Sampler preview, if we already have it, in an `Image Preview` is on that same Controller. 

 And if that's not possible, maybe an user option to disable all Sampler previews.

> (and maybe performance?)

Yeah, this is weird, because sometimes it's quite slow, but if I refresh the browser it's fast again.

**Comment by chrisgoringe** [comment]
Don't show previews in sampler if the image preview is in the same controller might be possible.

**Comment by chrisgoringe** [comment]
> Anyways, something happened to all the previews, now they are over the widgets (even in Load Image)

That's a change I made, so that the image is in the same place on all nodeblocks - I think it looks better?

**Comment by chrisgoringe** [comment]
This is getting way long, so I've added a few new issues to track bits and pieces...

#273 - don't show in sampler if showing in preview
#274 - preview node returns to interim image if you swap groups
#275 - images 'forget' their size

Closing this one as done since the basic functionality is there.

**Comment by JorgeR81** [comment]
> > Anyways, something happened to all the previews, now they are over the widgets (even in Load Image)
>
> That's a change I made, so that the image is in the same place on all nodeblocks - I think it looks better?

@LukeG89, I ask for that:  https://github.com/chrisgoringe/cg-controller/issues/263#issuecomment-2453387891

Now finally the `Load Image` is aligned with `Image Preview` !

I usually prefer to copy the "native" ComfyUI style, but not in this case.

I wanted this ( with nodes ) since I started using ComfyUI !

**Comment by JorgeR81** [comment]
I'm having some issues with the live previews:

- In this 3 stage workflow, the stage 1 preview is not shown in the big picture area.
It was working 2 day ago, when I took the screenshots above. 

![p1](https://github.com/user-attachments/assets/172564ef-9aae-478b-9256-76501c1004ab)

---

- Also, the live preview in stage 3 sampler ( Face Detailer ), sometimes gets wider during the generation. 
This was also not hapening before.

![prev1](https://github.com/user-attachments/assets/1f132535-29ec-4151-b76e-eb15f4dad5c8)

**Comment by JorgeR81** [comment]
**UPDATE:**

> - Also, the live preview in stage 3 sampler ( Face Detailer ), sometimes gets wider during the generation.

I don't seem to have this issue anymore

---

> - In this 3 stage workflow, the stage 1 preview is not shown in the big picture area.

In my workflow the big image preview is plugged to stage 3.  
I tried unplugging and plugging the wire and this seems to fix it !

Sometimes, it does not work. But if I plug it to stage 1 and then to stage 3, it also fixes the issue. 
It's not about changing the group configuration. It's just plugging and unplugging the wires that fixes the issue.  

But even if I save the "fixed" workflow, when I open it again it's also broken again.

**Comment by chrisgoringe** [comment]
Could you raise this as a new issue, with the workflow included? I'd prefer to keep feature issues and bugs separate where possible, when the bug isn't so serious as to render the feature unusable.

**Comment by JorgeR81** [comment]
> Could you raise this as a new issue, with the workflow included?

I've made a simpler 3 stage workflow, without custom nodes, but I was not able to replicate the bug.

I will keep testing this and will open an issue, when I have a better understanding of the cause of the issue, in the order workflow.

**Comment by chrisgoringe** [comment]
great. I'll leave this closed; we'll treat it as a bug if you reproduce.

**Comment by JorgeR81** [comment]
OK, I understand the issue.

This is caused by a switch I have in stage 2, to upscale the latent. ( I've only added it today ).

I don't think it's necessary to open an issue, since this is caused by this custom node ( a switch from the WAS node suite ).

A better solution would be to support the `Trung0246` switch in the Controller ( https://github.com/chrisgoringe/cg-controller/issues/283 )
Because it has more features and it seems to work better with the Controller.
It replaced the WAS switch, with `Trung0246` switch, in the workflow, the issue was also fixed. 
The problem with the `Trung0246` switch, is that it can't be added to the Controller.
